staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -104,6 +113,92 @@ type Options struct {
 
 	// now is used for testing. It defaults to time.Now.
 	now func() time.Time
+
+	// IssuersPerClaim supports distributed claims.
+	//
+	// See for details:
+	// http://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
+	//
+	// Lists, for each claim, the list of issuer URLs that are
+	// trusted to provide the respective distributed claim.  Only distributed
+	// claims with issuers as specified below will be resolved.  This is done
+	// to reduce the chance that an authentication request goes to an address
+	// that we don't trust to issue distributed claims.
+	//
+	// Multiple issuers are supported so that it would be possible to rotate
+	// the issuers over time (with sufficient advance notice) without
+	// disrupting cluster operations, while it is expected that more than one
+	// issuer per claim is present only if the distributed claims source change
+	// is imminent.
+	//
+	// Example:
+	// "groups" -> []string{"http://example1.com/foo","http://example2.com/bar"}
+	IssuersPerClaim map[string][]string
```

Comment:
why shouldn't the issuer decide rather than the k8s oidc client? what does this solve?

Reply:
I'm also for not adding new APIs.

Reply:
Two benefits, IMHO:
- Makes an explicit guarantee to the cluster owner that only issuers they explicitly approve of will ever be called to resolve distributed claims.
- Makes it possible for us to initialize the ID token provider ahead of time.

Reply:
@ericchiang could you please clarify your comment?  What new APIs are being added?

Reply:
> What new APIs are being added?

New flags.

> Makes an explicit guarantee to the cluster owner that only issuers they explicitly approve of will ever be called to resolve distributed claims. Note that it's not a decision by the oidc client; but a restriction on the acceptable values.

Users already put a lot of trust in their primary ID Token issuer. I feel like the benefit is small compared to adding more configuration nobs.

Reply:
@ericchiang , it's a fair concern.  Let me know what you decide.

Reply:
Thinking this this a little more, the only distributed claims we should allow are the username and groups claim. I would not expect a distributed claim to be able to provide the expiry, for example.

If we're worried about what remote addresses we can consult for this maybe we whitelist issuers?

```
--oidc-allowed-distributed-claims-issuer="https://auth.foo.com"
--oidc-allowed-distributed-claims-issuer="https://auth.bar.com"
```

However, when would an issuer provide a distributed claim that an admin wouldn't be comfortable with? If the admin is picking a username or group claim they're surely aware of if that claim can be distributed and where it'd be distributed to.

If admins are uncomfortable with this behavior maybe provide a flag to restore the old behavior once someone complains?

```
--oidc-disallow-distributed-claims=true
```

Reply:
> the only distributed claims we should allow are the username and groups claim

I agree.

> when would an issuer provide a distributed claim that an admin wouldn't be comfortable with? If the admin is picking a username or group claim they're surely aware of if that claim can be distributed and where it'd be distributed to.

I agree. If we support distributed claims, it seems odd to not default to using that support to obtain the info from the claims the administrator explicitly configured.

Reply:
> Thinking this this a little more, the only distributed claims we should allow are the username
> and groups claim (edit: or maybe even only allow it for the groups claim). 

Could you explain why you think this is the case?  (e.g. for example, only nonstandard claims are allowed to be distributed?)

> I would not expect a distributed claim to be able to provide the expiry, for example.

Hm, I have a slightly different view, though I'm not an expert in the matter.  It seems that a distributed claim response *must* provide the expiry.

Otherwise, it becomes similar to a long-lived credential, which could be stolen and replayed.  With a timestamp, whatever is examining the credential has a way to determine if the credential is fresh or not and therefore if it can believe the claims. This seems like quite an important requirement from a security perspective.

> If we're worried about what remote addresses we can consult for this maybe we whitelist issuers?

That would work.

> However, when would an issuer provide a distributed claim that an admin wouldn't be comfortable with? If the admin is picking a username or group claim they're surely aware of if that claim can be distributed and where it'd be distributed to.

I think from an auditability perspective it is much stronger to claim "we know exactly what calls our API server is making" than "our API server may be redirected to wherever our IDP says".

But if you think that's not a big concern, I'll be happy to oblige.

Reply:
>> I would not expect a distributed claim to be able to provide the expiry, for example.
>
> Hm, I have a slightly different view, though I'm not an expert in the matter. It seems that a distributed claim response must provide the expiry.

I think you're talking about two different things. First, we verify the presented id_token, including its `iss`, `exp`, `nbf` claims. *That* is the `exp` claim Eric was saying he wouldn't expect to be able to be distributed.

Once verified, we look for the username and group claims. Those could be distributed, and if they are, according to the spec, the response from the distributed endpoint is *another* JWT that must be verified (including `iss`, and potentially its own `exp`, and `nbf` claims)

Reply:
> First, we verify the presented id_token, including its iss, exp, nbf claims. That is the exp claim Eric was saying he wouldn't expect to be able to be distributed.

Ah I see.  Indeed, not all claims may be named in `_claim_names`.   In this implementation, a normal claim always wins if there is a claim name clash, which means `exp` will never be resolved through `_claim_sources`.  But, let's make that statement even stronger, then.

To summarize my understanding of what would be acceptable to you:

- Only allow `groups` claim to be distributed (that is, whatever the value has been set in the OIDC configuration to mean groups).
- Require admins to specify `--oidc-allowed-distributed-claim-issuer` to list all acceptable issuers.
- Introduce `--oidc-disallow-distributed-claims` as an escape hatch.

What do you think?

Reply:
> Only allow groups claim to be distributed (that is, whatever the value has been set in the OIDC configuration to mean groups).

+1

> Require admins to specify --oidc-allowed-distributed-claims-issuer to list all acceptable issuers.

-1

I still think that admins are actually more tolerant of requests to different sites than you expect. `accounts.google.com` returns a keys URL hosted under `www.googleapis.com`, for example.

> Introduce --oidc-disallow-distributed-claims as an escape hatch.

I'd like to see someone request this behavior first.

Reply:
OK, I scaled back the implementation to support only groups claim resolution through distributed claims.  Also removed all flags and options.  Better?